runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java

```diff
@@ -187,6 +186,44 @@ private MonitoringInfo counterUpdateToMonitoringInfo(MetricUpdate<Long> metricUp
     return builder.build();
   }
 
+  /**
+   * @param metricUpdate
+   * @return The MonitoringInfo generated from the metricUpdate.
+   */
+  @Nullable
+  private MonitoringInfo distributionUpdateToMonitoringInfo(
+      MetricUpdate<org.apache.beam.runners.core.metrics.DistributionData> metricUpdate) {
+    SimpleMonitoringInfoBuilder builder = new SimpleMonitoringInfoBuilder(true);
+    MetricName metricName = metricUpdate.getKey().metricName();
+    if (metricName instanceof MonitoringInfoMetricName) {
+      MonitoringInfoMetricName monitoringInfoName = (MonitoringInfoMetricName) metricName;
+      // Represents a specific MonitoringInfo for a specific URN.
+      builder.setUrn(monitoringInfoName.getUrn());
+      for (Entry<String, String> e : monitoringInfoName.getLabels().entrySet()) {
+        builder.setLabel(e.getKey(), e.getValue());
+      }
+    } else { // Note: (metricName instanceof MetricName) is always True.
+      // Represents a user counter.
+      builder
+          .setUrn(MonitoringInfoConstants.Urns.USER_DISTRIBUTION_COUNTER)
+          .setLabel(
+              MonitoringInfoConstants.Labels.NAMESPACE,
+              metricUpdate.getKey().metricName().getNamespace())
+          .setLabel(
+              MonitoringInfoConstants.Labels.NAME, metricUpdate.getKey().metricName().getName());
+
+      // Drop if the stepname is not set. All user counters must be
+      // defined for a PTransform. They must be defined on a container bound to a step.
+      if (this.stepName == null) { // TODO rethink this/rm this check.
+        return null;
```

Comment:
Should we log a warning here? This would reflect an error right?

I understand that we may not want to surface these warnings to users... how should we surface warnings to Beam developers and not to Beam users? (fair to answer this question later : ))

Reply:
added todo